base_guicontrol.cpp

```diff
@@ -51,3 +51,59 @@ int MainFrame::getDepth(){
 double MainFrame::getProbeRadius(){
   return r_probe;
 }
+
+void MainFrame::generateAtomList(std::vector<std::tuple<std::string, int, double>>& symbol_number_radius){
+  // Clear previous instance of the atom list grid
+  // DeleteRows causes an error if there is 0 row!
+  if (atomListGrid->GetNumberRows() > 0) {
+    atomListGrid->ClearGrid();
+    atomListGrid->DeleteRows(0,atomListGrid->GetNumberRows(),true);
+  }
+
+  for (int i = 0; i<symbol_number_radius.size(); i++) {
+    atomListGrid->AppendRows(1, true);
+    atomListGrid->SetCellValue(std::get<0>(symbol_number_radius[i]),i,1);
+    atomListGrid->SetCellValue(std::to_string(std::get<1>(symbol_number_radius[i])),i,2);
+    atomListGrid->SetCellValue(std::to_string(std::get<2>(symbol_number_radius[i])),i,3);
+    atomListGrid->SetReadOnly(i,1,true);
+    atomListGrid->SetReadOnly(i,2,true);
+    if (std::wcstod(atomListGrid->GetCellValue(i,3), NULL) == 0){
+      atomListGrid->SetCellBackgroundColour(i,1,col_grey_cell);
+      atomListGrid->SetCellBackgroundColour(i,2,col_grey_cell);
+      atomListGrid->SetCellBackgroundColour(i,3,col_red_cell);
+    }
+    else {
+      atomListGrid->SetCellValue("1",i,0);
+    }
+  }
+  atomListGrid->Refresh();
+  atomListGrid->Fit();
+  FitSizes();
+}
+
+std::string MainFrame::generateChemicalFormulaFromGrid(){
+// TODO: For now, the chemical formula is generated in alphabetical order
+// however, the convention is to start with C, then H then by alphabetical order
+// we could modify the function to follow the convention
+  std::string chemical_formula = "";
+  for (int i = 0; i < atomListGrid->GetNumberRows(); i++){
+    if (atomListGrid->GetCellValue(i,0) == "1"){
+      std::string symbol = atomListGrid->GetCellValue(i,1).wxString::ToStdString();
+      std::string number = atomListGrid->GetCellValue(i,2).wxString::ToStdString();
+      chemical_formula.append(symbol);
+      chemical_formula.append(number);
+    }
+  }
+  return chemical_formula;
+}

```

Comment:
Is there a reason why you generate the chemical formula from the grid rather than get it directly from the input file, or from the tuple that contains all information on the input file? I guess maybe so you don't have to recalculate the numbers? Still, it seems strange to me to get information through such an indirect way.

Reply:
The chemical formula is generated based on the elements included by the user. This information is only available in the grid.

Reply:
Got it!